client/src/components/Routers.js

```diff
@@ -39,13 +41,18 @@ export default function Routers() {
                     />
                 ))
               }
-              <Route>
-                <Redirect to="/login" />
-              </Route>
+              <Route
+                render={({location}) => (
+                  <Redirect to={{
+                    pathname: appContext.isLoggedIn() ? "/" : "/login",
+                    state: { from: location }
+                  }}/>
+                )}
+              />
             </Switch>
           </Grid>
         </Grid>
       </Grid>
     )
-  }, [appContext.routes])
```

Comment:
Why change this? Namely, the routes object is memoized so that we don't waste time recomputing the routes if the item doesn't change. If the user changes (via deep equals) so does the routes, but the current implementation may set the same user object a few times, causing the context object to be regenerated. But the pointer to context.routes will be the same.

If the permissions change, the user changes, and routes changes with it, afaik

Reply:
The additional check for `appContext.isLoggedIn()` to decide between `/` and `/login` triggered a React dependency warning suggesting `appContext` needed to be included as a dependency (react-hooks/exhaustive-deps).

I assume this is React saying that it can't know what state changes in `appContext` would cause changes in the value returned by `appContext.isLoggedIn()`.

I'll change `isLoggedIn` to be a variable rather than a method and include it directly as a dependency.